### ISSUE TITLE
Incorrect path passing for LSP server

### ISSUE THREAD

**Issue by to266** [helix-editor/helix#256]
### Reproduction steps

 - Create (in my case rust) project with the name appearing somewhere higher in the path tree. e.g. `/home/to266/projects/funny/some/folder/workspace/funny`. If `workspace` literally corresponds to rust workspace, and `funny` inside is a rust package withing the workspace.
 - Start `hx` inside `funny` the package
 - Observe that LSP is not initialized. The logs show
```
2021-06-14T09:40:14.641 helix_lsp::transport [ERROR] err <- [ERROR rust_analyzer] failed to find any projects in [AbsPathBuf("/home/to266/projects/funny")]
2021-06-14T09:40:14.642 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/showMessage","params":{"type":1,"message":"rust-analyzer failed to discover workspace"}}

```

### Environment

- Platform: Linux
- Helix version: 0.2.0

**Comment by archseer** [comment]
I think this is a rust-analyzer issue: https://github.com/rust-analyzer/rust-analyzer/issues/1798

We try to detect a `.git` repository root in any of the parent directories, if that fails we fall back to passing `None` to the LSP and letting it figure out the correct setting.

**Comment by nrdxp** [comment]
It's not just rust-analyzer, as rls failed with the same issue.

**Comment by archseer** [comment]
Yeah currently we don't consume those roots yet. I think we need to change find_roots to accept a `markers` list, then modify this: https://github.com/helix-editor/helix/blob/1bda4541491bf0fc8129214bb52515a32b8d4978/helix-core/src/lib.rs#L42-L46

The reason I haven't done this yet is because I don't know what the behavior should be here. If we stop at the first parent dir with a `Cargo.toml` we might miss the fact that we're in a workspace. On the other hand, if this is a multi-language repository, searching from the root -> cwd and finding `.git` might not be what we want (instead we want the subfolder with `Cargo.toml`)

**Comment by pickfire** [comment]
I think if we look at `.git` at home we might hit `$HOME` as our root which may not be what the user want.

**Comment by archseer** [comment]
$HOME doesn't contain `.git` though, it contains a `.gitconfig`/`.gitignore`. I think likely the correct way to do this is to scan for a git repo first, then scan for defined roots next.

**Comment by pickfire** [comment]
> $HOME doesn't contain .git though, it contains a .gitconfig/.gitignore. I think likely the correct way to do this is to scan for a git repo first, then scan for defined roots next.

I used to have it until it broke some applications so I use an alias `home` `git --work-tree=/home/ivan --git-dir=/home/ivan/.home`, so we may want to beware of it. Technically it is possible to have `.git/` it home for say dotfiles so we may want to avoid it.

**Comment by nrdxp** [comment]
Perhaps, at least for rust, we could search for a directory containing both a `Cargo.toml` and a `.git` directory. Seems like a fairly clear indication of the top level rust crate.

**Comment by to266** [comment]
> Perhaps, at least for rust, we could search for a directory containing both a `Cargo.toml` and a `.git` directory. Seems like a fairly clear indication of the top level rust crate.

Perhaps for rust-only projects. In my original example it's not the case
```
/home/
    to266/
        projects/
            funny/                       # git root
                some/                # just folder
                    folder/          # just folder
                        workspace/       # rust workspace
                            funny        # Crate
```
To me it seems the proper solution is actually reading the `Cargo.toml` files. If it contains the keyword `[workspace]`, then it's a workspace, and no other (at least related) will be above that (because nested workspaces are not a thing https://github.com/rust-lang/cargo/issues/5042).

if additional verification is wanted, then of course it's possible to check that the `Cargo.toml` in workspace contains the member crate, which is presumably where we started in the first place.

**Comment by archseer** [comment]
nvim-lsp uses `cargo metadata --no-deps --format-version 1` to dump data then search for `workspace_root`. Wish it was more straightforward.

**Comment by matoous** [comment]
I think this is still an issue, and not just for Julia but any mono-repo. Our structure looks like this

```
repo/
    .git
    apps/
        ui/
            package.json
        backend/
            go.mod
```

`gopls` in the `repo/apps/backend` doesn't work either currently.

I think (as mentioned above) the issue is with the `find_root` function https://github.com/helix-editor/helix/blob/3156577fbf1a97e07e90e11b51c66155f122c3b7/helix-core/src/lib.rs#L41. The question is what should be the optimal behavior:

* still look for root where the `.git` is but initialize `lsp` in directory that matches `roots`.
* look for `roots` or `.git`
* always keep the current folder as the root and only initialize the LSP in the first parent directory that matches `roots` or stop at `.git`
* something else?

If I understand it correctly, ideally this call to `find_root` https://github.com/helix-editor/helix/blob/master/helix-lsp/src/client.rs#L228 would take the `roots` configuration into consideration.

**Comment by archseer** [comment]
I think the procedure should be:

1) look for the highest directory with the marker file (we have these in languages l.toml, they're just not used currently)
2) else look for a .git folder
3) fallback to using current directory

**Comment by amousset** [comment]
I opened https://github.com/helix-editor/helix/pull/1370 with a slightly different logic (stop looking for markers when reaching git repo boundary, but not sure it makes more sense)

**Comment by archseer** [comment]
Resolved by #1370
